File: scripts/gen_favicons.py
#!/usr/bin/env python3
"""Downsize assets/img/icon-512.png (from gen_favicons.mjs) to the full favicon
set: icon-192, favicon-96, favicon-48, favicon-32 + a multi-size favicon.ico at
the web root (SERP favicon needs >=48px).
  node scripts/gen_favicons.mjs && python3 scripts/gen_favicons.py
"""
import os
from PIL import Image

ROOT = os.path.join(os.path.dirname(__file__), "..")
IMG = os.path.join(ROOT, "assets", "img")
src = Image.open(os.path.join(IMG, "icon-512.png")).convert("RGBA")

for size, name in [(192, "icon-192.png"), (96, "favicon-96.png"),
                   (48, "favicon-48.png"), (32, "favicon-32.png")]:
    src.resize((size, size), Image.LANCZOS).save(os.path.join(IMG, name))
    print("wrote assets/img/" + name)

src.resize((48, 48), Image.LANCZOS).save(
    os.path.join(ROOT, "favicon.ico"), sizes=[(16, 16), (32, 32), (48, 48)])
print("wrote favicon.ico (16/32/48)")

File: scripts/gen_favicons.mjs
// gen_favicons.mjs — rasterize assets/img/favicon.svg to a 512px PNG via the
// globally-installed Playwright Chromium (no network). Pillow then downsizes
// (scripts/gen_favicons.py) to 192/96/48/32 + favicon.ico.
//   node scripts/gen_favicons.mjs && python3 scripts/gen_favicons.py
import { createRequire } from 'module';
const require = createRequire('/opt/node22/lib/node_modules/');
const { chromium } = require('playwright');
import { readFileSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const svg = readFileSync(resolve(ROOT, 'assets/img/favicon.svg'), 'utf8');
const html = `<!doctype html><html><head><style>*{margin:0}html,body{width:512px;height:512px;background:transparent}svg{width:512px;height:512px;display:block}</style></head><body>${svg}</body></html>`;

const browser = await chromium.launch();
const page = await browser.newPage({ viewport: { width: 512, height: 512 } });
await page.setContent(html, { waitUntil: 'load' });
await page.screenshot({ path: resolve(ROOT, 'assets/img/icon-512.png'), omitBackground: true });
await browser.close();
console.log('✓ assets/img/icon-512.png (512x512)');
